odoo/models.py

```diff
@@ -5590,35 +5590,66 @@ def exists(self):
         valid_ids = set([r[0] for r in self._cr.fetchall()] + new_ids)
         return self.browse(i for i in self._ids if i in valid_ids)
 
-    def _check_recursion(self, parent=None):
+    def _check_recursion(self, field_name=None):
         """
-        Verifies that there is no loop in a hierarchical structure of records,
-        by following the parent relationship using the **parent** field until a
-        loop is detected or until a top-level record is found.
+        Verifies that there is no loop in a graph of records,
+        by following a given relationship.
+        By default the **parent** field is used as the relationship.
 
-        :param parent: optional parent field name (default: ``self._parent_name``)
+        :param field_name: optional field name (default: ``self._parent_name``)
         :return: **True** if no loop was found, **False** otherwise.
         """
-        if not parent:
-            parent = self._parent_name
+        if not field_name:
+            field_name = self._parent_name
+        field = self._fields.get(field_name)
+        if not field:
+            raise ValueError('invalid field_name: %r' % (field_name,))
+        if not (
+            field.type in ('many2many', 'many2one')
+            and field.comodel_name == self._name
+            and field.store
+        ):
+            raise ValueError('field must be a relation on itself: %r' % (field_name,))
```

Comment:
```suggestion
            raise ValueError('field must be a many2one or many2many relation on itself: %r' % (field_name,))
```
Make sure the code and error/comments are consistent. We actually can check one2many with little code diff. But I think it is not very usefulm, unless we need to check the loop in non-constraints methods.

Reply:
Updated. I think it is not really useful to check one2many, in the implementation; we would need to get the inverse name, but the semantics of ids to check would be harder to define.